### DISCUSSION TITLE
[Feature Request]: Add i18n

### DISCUSSION THREAD

**Discussion by Fghurayri** [blitz-js/blitz#1776] [Ideas]
### What do you want and why?

i18n is immensely needed in any web __framework__. 

### Possible implementation(s)

The Nextjs community are yet to agree on the best possible way to implement i18n while keeping SSR / SSG / Serverless working as expected.

Can this be solved by Blitz?

### Additional context

https://github.com/zeit/next.js/discussions/10651

**Comment by flybayer** [comment]
I've never implemented i18n myself, but I agree Blitz should solve this.

So I'd love to hear proposals or ideas on how we should do this!!

**Reply by MrLeebo** [reply]
I converted this issue into a discussion because this would be a great feature to have, but it sounds like possible solutions are coming from all different directions. We need to suss out how to best deliver this feature in keeping with i18n standards and Blitz's goals.

**Comment by waptik** [comment]
So after going through all possible solutions/examples, [with-i18n-rosetta](https://github.com/zeit/next.js/tree/canary/examples/with-i18n-rosetta) looks like the best possible solution at the moment ([demo](https://with-i18n-rosetta.now.sh/)) . No difficult setup nor extra build steps.
Besides, you can also use your locales as packages and easily import them in your projects like how [Uppy](https://uppy.io/docs/uppy/#locale) does it.

**Comment by flybayer** [comment]
Awesome, thank you @waptik.

So it appears very likely that the Next.js team is working on an official i18n solution ([source](https://twitter.com/timneutkens/status/1264653634639409153) and [source](https://twitter.com/timneutkens/status/1294642999096496129)). I think we should wait and see that before doing something ourselves.

Only bummer is they haven't given a timeline.

**Comment by waptik** [comment]
@flybayer 😱 wow, that's great news 😋. I think an official solution is the best way to go, especially if it's coming from the authors.
So it's worth waiting 😊
Thanks for sharing the good news

**Comment by wescopeland** [comment]
A quick update on this: [next-i18next now supports serverless i18n from their beta branch](https://github.com/vercel/next.js/discussions/10651#discussioncomment-34992). The implementation relies on path rewrites, which are currently behind an experimental flag but [will be GA on the next stable release](https://github.com/vercel/next.js/pull/14978#issuecomment-659761936).

**Comment by flybayer** [comment]
The [Next.js RFC on i18n Routing](https://github.com/vercel/next.js/discussions/17078) has just been posted!

**Comment by jan-wilhelm** [comment]
any news on this, now that NextJS 10 is used in Blitz and includes official support for i18n routing?

**Comment by flybayer** [comment]
@jan-wilhelm no news! Now that that's released, it's up to y'all to decide if there's more stuff that we should add on top of that in blitz.

**Comment by jan-wilhelm** [comment]
After working with "raw" NextJS i18n for a while, I've found [`next-translate`](https://github.com/vinissimus/next-translate) to be a small and very intuitive library for i18n. I would suggest using this for Blitz too, but I am not sure if it works with Blitz's way of moving NextJS `pages` around. Maybe somebody can try that out? I'm quite busy these days, but could help with that sometime soon.

**Reply by boredland** [reply]
I gave that a try, but failed. There's two locations I know of, where next-translate searches for i18n in a static path:
https://github.com/vinissimus/next-translate/blob/master/src/plugin/index.ts#L13
https://github.com/vinissimus/next-translate/blob/master/src/plugin/index.ts#L54

As a first try I just extended that to search inside the app-dir:
https://github.com/finalfinalphotos/next-translate/commit/939b818a005c7f3af07c1ebf39e4f5933d6f2c86

But yeah, that gave me some weird rule-of-hooks validation I didn't understand.

Has anybody already got a working i18n setup yet?

**Reply by BjoernRave** [reply]
I created an issue on the `next-translate` repo: https://github.com/vinissimus/next-translate/issues/499

**Comment by devtombiz** [comment]
Any news about this issue ? Do we have a working example of translations in blitz ?

**Reply by flybayer** [reply]
No news on our side, but latest next-translate should support Blitz. We still have [this issue](https://github.com/blitz-js/blitz/issues/2680) which likely affects some translation libraries.

If anyone has a working example, please PR it to our examples folder!

**Reply by devtombiz** [reply]
Hi Brandon,
I made a localization example with next-rosetta. It seems to work fine. You can see my PR here: https://github.com/devtombiz/blitzI18nRosetta/pull/1/files
If this is satisfactory to you, then I can create a PR on one of your examples folder (which one ?)
-Tom

**Reply by flybayer** [reply]
@devtombiz awesome!! Please add to the top level examples/ folder

**Reply by devtombiz** [reply]
Done, see PR https://github.com/blitz-js/blitz/pull/2872
